Sources/Fuzzilli/FuzzIL/AbstractInterpreter.swift

```diff
@@ -36,22 +36,48 @@ public struct AbstractInterpreter {
     
     public mutating func reset() {
         ifStack.removeAll()
-        stack = [VariableMap<Type>()]
+        stateDiffsStack = [VariableMap<Type>()]
         propertyTypes.removeAll()
         methodSignatures.removeAll()
     }
 
-    /// Get variables with new type which have different type in old state
-    /// Used after block end when some state on stack is removed
-    private func getStateChanges(oldState: VariableMap<Type>, newState: VariableMap<Type>) -> [(Variable, Type)] {
-        var typeChanges = [(Variable, Type)]()
-        for (variable, type) in newState {
-            let oldType = oldState[variable]
-            if oldType == nil || type != oldType  {
-                typeChanges.append((variable, type))
+    /// Update current variable types and type diffs
+    /// Used after block end when some state is removed
+    private mutating func updateStateDiffs(
+        lastStateDiff: VariableMap<Type>, typeChanges: inout [(Variable, Type)], mergeTypes: Bool = true
+    ) {
+        for (variable, oldType) in lastStateDiff {
+            var newType = oldType
+
+            if mergeTypes {
+                // If variable did not exist in parent state we keep its current type
+                // So we won't record any new type changes
+                if oldType == .nothing {
```

Comment:
I think it would be good to explain your trick with `.nothing` expressing that a variable wasn't defined in the outer scope somewhere. I guess either here or in `setType` where you set it